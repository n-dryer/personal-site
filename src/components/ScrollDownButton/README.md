# ScrollDownButton Component

This directory contains the `ScrollDownButton` component, its tests, and related files.

## Purpose

The `ScrollDownButton` component provides a button that initially prompts the user to scroll down to view more content (e.g., the timeline section).
It appears with an animation and then disappears after the first scroll event or when clicked, and remains hidden thereafter.

## Usage

```tsx
import { ScrollDownButton } from '@/components/ScrollDownButton'; // Or the correct relative path

<ScrollDownButton />
```

## Props

This component does not accept any props. It manages its own visibility based on scroll events and relies on DOM element IDs (e.g., scrolling to `#timeline`) for its primary action. 