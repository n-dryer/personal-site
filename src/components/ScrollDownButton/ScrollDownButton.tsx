// ◀︎ LLM-modified - Enhanced responsive sizing using clamp() functions for better mobile/desktop scaling
import React, { useState, useEffect, useCallback } from 'react';
import { motion, Variants } from 'framer-motion';
import { ChevronDown } from 'lucide-react';

/**
 * ScrollDownButtonComponent displays a button that scrolls the user to the timeline section.
 * It appears with an animation and has a continuous subtle animation to draw attention.
 * The button hides itself once scrolled or if the user scrolls close to the bottom of the page.
 * This component is memoized for performance optimization, although it takes no props,
 * its internal logic and effects are considered non-trivial.
 *
 * @returns {React.ReactElement | null} The rendered ScrollDownButton, or null if not visible.
 * @see React.memo
 */
const ScrollDownButtonComponent = (): React.ReactElement | null => {
  const [isVisible, setIsVisible] = useState(true);
  const [animationPhase, setAnimationPhase] = useState<'pulse' | 'bounce'>('pulse');

  const handleHideAndRemoveListener = useCallback(() => {
    setIsVisible(false);
    window.removeEventListener('scroll', handleHideAndRemoveListener);
  }, []);

  useEffect(() => {
    // Add the event listener when the component mounts
    window.addEventListener('scroll', handleHideAndRemoveListener);

    // Return a cleanup function to remove the event listener when the component unmounts
    // or if handleHideAndRemoveListener removes itself.
    return () => {
      window.removeEventListener('scroll', handleHideAndRemoveListener);
    };
  }, [handleHideAndRemoveListener]);

  const handleScrollToTimeline = useCallback((): void => {
    handleHideAndRemoveListener(); // Hide button on click and remove listener
    const timelineElement = document.getElementById('timeline') ||
      document.querySelector('[data-section="timeline"]') ||
      document.querySelector('main');

    if (timelineElement) {
      timelineElement.scrollIntoView({
        behavior: 'smooth',
        block: 'start'
      });
    }
  }, [handleHideAndRemoveListener]);

  const buttonVariants: Variants = {
    bounce: {
      y: [0, -1.5, 0], // More visible bounce animation
      transition: {
        duration: 2.0, // 2s bounce duration
        repeat: Infinity,
        repeatType: 'loop' as const,
        ease: 'easeInOut'
      }
    },
    pulse: {
      scale: [1, 1.02, 1],
      transition: {
        duration: 3.0,
        repeat: Infinity,
        repeatType: 'loop' as const,
        ease: 'easeInOut'
      }
    },
    hover: {
      scale: 1.05,
      y: -1,
      transition: {
        type: 'spring',
        stiffness: 300,
        damping: 20
      }
    },
    tap: {
      scale: 0.95,
      y: 0,
      transition: { duration: 0.15 }
    }
  };

  const [prefersReducedMotion, setPrefersReducedMotion] = useState<boolean>(false);

  useEffect(() => {
    if (typeof window !== 'undefined' && window.matchMedia) {
      const mediaQuery = window.matchMedia('(prefers-reduced-motion: reduce)');
      setPrefersReducedMotion(mediaQuery.matches);

      const handleChange = (e: MediaQueryListEvent) => {
        setPrefersReducedMotion(e.matches);
      };

      mediaQuery.addEventListener('change', handleChange);
      return () => mediaQuery.removeEventListener('change', handleChange);
    }
  }, []);

  // Switch to bounce after a delay
  useEffect(() => {
    const timer = setTimeout(() => {
      setAnimationPhase('bounce');
    }, 1750);
    return () => clearTimeout(timer);
  }, []);

  if (!isVisible) {
    return null;
  }

  return (
    <motion.div
      className="fixed bottom-6 left-1/2 z-30 -translate-x-1/2 transform"
      initial={{ opacity: 0, scale: 0.4, y: 16 }}
      animate={{ opacity: 1, scale: 1, y: 0 }}
      transition={{
        opacity: { duration: 1.0, ease: 'easeOut' },
        scale: { type: 'spring', stiffness: 300, damping: 20, duration: 0.5 },
        y: { type: 'spring', stiffness: 300, damping: 20, duration: 1.0 }
      }}
    >
      <motion.button
        className="focus-visible:ring-accent/40 flex items-center justify-center rounded-full bg-accent text-on-accent shadow-lg focus:outline-none focus-visible:ring-2"
        style={{
          width: 'clamp(2.75rem, 4vw, 3.5rem)',
          height: 'clamp(2.75rem, 4vw, 3.5rem)',
        }}
        onClick={handleScrollToTimeline}
        whileHover="hover"
        whileTap="tap"
        variants={buttonVariants}
        animate={prefersReducedMotion ? undefined : animationPhase}
        aria-label="Scroll to timeline section"
        role="button"
        tabIndex={0}
        onKeyDown={(e) => {
          if (e.key === 'Enter' || e.key === ' ') {
            e.preventDefault();
            handleScrollToTimeline();
          }
        }}
      >
        <ChevronDown
          style={{
            width: 'clamp(1rem, 2.5vw, 1.5rem)',
            height: 'clamp(1rem, 2.5vw, 1.5rem)',
          }}
        />
      </motion.button>
    </motion.div>
  );
};

/**
 * Memoized ScrollDownButton component.
 * @see ScrollDownButtonComponent
 */
export const ScrollDownButton = React.memo(ScrollDownButtonComponent); 