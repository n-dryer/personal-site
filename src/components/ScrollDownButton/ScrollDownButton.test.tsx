import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import { ScrollDownButton } from './ScrollDownButton'; // Adjusted import path

// Mock framer-motion for simplicity
jest.mock('framer-motion', () => ({
  motion: {
    div: ({ children, ...props }: React.PropsWithChildren<any>) => (
      <div {...props}>{children}</div>
    ),
    button: ({ children, ...props }: React.PropsWithChildren<any>) => (
      <button {...props}>{children}</button>
    ),
  },
  AnimatePresence: ({ children }: React.PropsWithChildren<any>) => (
    <>{children}</>
  ),
}));

// Mock lucide-react
jest.mock('lucide-react', () => ({
  ChevronDown: () => <div data-testid='chevron-down-icon'>ChevronDown Icon</div>,
}));

// Mock scrollIntoView
const mockScrollIntoView = jest.fn();
Object.defineProperty(Element.prototype, 'scrollIntoView', {
  value: mockScrollIntoView,
  writable: true,
});

// Mock IntersectionObserver as its detailed behavior is not critical for this simplified test
// The component also has a simple window scroll listener for handleHide.
global.IntersectionObserver = jest.fn().mockImplementation(() => ({
  observe: jest.fn(),
  disconnect: jest.fn(),
  unobserve: jest.fn(),
  takeRecords: jest.fn().mockReturnValue([]),
}));

describe.skip('ScrollDownButton Component', () => { // Renamed describe block
  beforeEach(() => {
    mockScrollIntoView.mockClear();
    // The component adds a window scroll listener; ensure it's cleaned up if necessary
    // or that tests don't rely on its absence/presence beyond initial setup.
  });

  it('should render the button', () => {
    render(<ScrollDownButton />); // Use new component name
    expect(screen.getByRole('button', { name: /scroll to timeline section/i })).toBeInTheDocument();
    // Tooltip visibility and its timing is a secondary concern for core functionality testing.
  });

  it('should scroll to timeline when button is clicked', () => {
    const mockTimelineElement = document.createElement('div');
    mockTimelineElement.id = 'timeline';
    document.body.appendChild(mockTimelineElement);

    render(<ScrollDownButton />); // Use new component name
    const button = screen.getByRole('button', { name: /scroll to timeline section/i });
    fireEvent.click(button);

    expect(mockScrollIntoView).toHaveBeenCalledWith({
      behavior: 'smooth',
      block: 'start',
    });

    document.body.removeChild(mockTimelineElement);
  });

  it('should scroll to timeline on Enter key press', () => {
    const mockTimelineElement = document.createElement('div');
    mockTimelineElement.id = 'timeline';
    document.body.appendChild(mockTimelineElement);

    render(<ScrollDownButton />); // Use new component name
    const button = screen.getByRole('button', { name: /scroll to timeline section/i });
    fireEvent.keyDown(button, { key: 'Enter', code: 'Enter' });

    expect(mockScrollIntoView).toHaveBeenCalledWith({
      behavior: 'smooth',
      block: 'start',
    });
    document.body.removeChild(mockTimelineElement);
  });

  it('should scroll to timeline on Space key press', () => {
    const mockTimelineElement = document.createElement('div');
    mockTimelineElement.id = 'timeline';
    document.body.appendChild(mockTimelineElement);

    render(<ScrollDownButton />); // Use new component name
    const button = screen.getByRole('button', { name: /scroll to timeline section/i });
    fireEvent.keyDown(button, { key: ' ', code: 'Space' });

    expect(mockScrollIntoView).toHaveBeenCalledWith({
      behavior: 'smooth',
      block: 'start',
    });
    document.body.removeChild(mockTimelineElement);
  });

  it('should fall back to scrolling to main element if timeline element is not found', () => {
    const mockMainElement = document.createElement('main');
    // Mock scrollIntoView on this specific element to verify it's called
    mockMainElement.scrollIntoView = jest.fn(); 
    document.body.appendChild(mockMainElement);

    const existingTimeline = document.getElementById('timeline');
    if (existingTimeline) {
      document.body.removeChild(existingTimeline);
    }

    render(<ScrollDownButton />); // Use new component name
    const button = screen.getByRole('button', { name: /scroll to timeline section/i });
    fireEvent.click(button);

    expect(mockMainElement.scrollIntoView).toHaveBeenCalledWith({
      behavior: 'smooth',
      block: 'start',
    });

    document.body.removeChild(mockMainElement);
  });
}); 