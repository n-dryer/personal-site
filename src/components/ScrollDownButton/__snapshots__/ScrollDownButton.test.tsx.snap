// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScrollDownButton Component renders correctly 1`] = `
<div>
  <div
    class="fixed bottom-6 right-6 z-40"
  >
    <div
      class="absolute bottom-full right-0 mb-4 rounded-lg bg-white px-3 py-2 text-sm text-gray-700 shadow-md dark:bg-gray-800 dark:text-gray-200"
      data-testid="motion-div"
    >
      Scroll to timeline
    </div>
    <button
      aria-label="Scroll to timeline"
      class="flex h-14 w-14 items-center justify-center rounded-full bg-gradient-to-r from-blue-500 to-blue-600 text-white shadow-lg transition-colors duration-300 hover:from-blue-600 hover:to-blue-700 focus:outline-none focus-visible:ring-2 focus-visible:ring-blue-500 focus-visible:ring-offset-2 dark:focus-visible:ring-offset-gray-900"
      data-testid="motion-button"
    >
      <div
        data-testid="chevron-down-icon"
      >
        ChevronDown Icon
      </div>
    </button>
  </div>
</div>
`;
